presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -24,69 +24,117 @@
         this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
         this.objectMapper = requireNonNull(objectMapper, "handleResolver is null");
+
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
+                .maximumSize(CACHE_MAXIMUM_SIZE)
+                .expireAfterWrite(CACHE_EXPIRES_AFTER_WRITE_MILLIS.toMillis(), MILLISECONDS);
+
+        this.catalogExistsCache = cacheBuilder.build(CacheLoader.from(this::loadCatalogExists));
+        this.schemaExistsCache = cacheBuilder.build(CacheLoader.from(this::loadSchemaExists));
+        this.listSchemaNamesCache = cacheBuilder.build(CacheLoader.from(this::loadListSchemaNames));
+        this.getTableHandleCache = cacheBuilder.build(CacheLoader.from(this::loadGetTableHandle));
+        this.listTablesCache = cacheBuilder.build(CacheLoader.from(this::loadListTables));
+        this.listViewsCache = cacheBuilder.build(CacheLoader.from(this::loadListViews));
+        this.getViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetViews));
+        this.getViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetView));
+        this.getMaterializedViewCache = cacheBuilder.build(CacheLoader.from(this::loadGetMaterializedView));
+        this.getReferencedMaterializedViewsCache = cacheBuilder.build(CacheLoader.from(this::loadGetReferencedMaterializedViews));
     }
 
+    /*
+        Metadata Manager Methods
+     */
+
     @ThriftMethod
     public boolean schemaExists(TransactionInfo transactionInfo, SessionRepresentation session, CatalogSchemaName schema)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.schemaExists(session.toSession(sessionPropertyManager), schema);
+        return get(schemaExistsCache, new CacheKey<>(session, schema));
     }
 
     @ThriftMethod
     public boolean catalogExists(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.catalogExists(session.toSession(sessionPropertyManager), catalogName);
+        return get(catalogExistsCache, new CacheKey(session, catalogName));
     }
 
     @ThriftMethod
     public String listSchemaNames(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        List<String> schemaNames = metadataProvider.listSchemaNames(session.toSession(sessionPropertyManager), catalogName);
-        if (!schemaNames.isEmpty()) {
-            return writeValueAsString(schemaNames, objectMapper);
-        }
-        return EMPTY_STRING;
+        List<String> schemaNames = get(listSchemaNamesCache, new CacheKey(session, catalogName));
+        return schemaNames.isEmpty()
+                ? EMPTY_STRING
```

Comment:
Actually, orthogonal from thrift or json, why can't we just directly `writeValueAsString` and we need an empty string?

Reply:
So since some empty results are represented as an Optional.empty(), we pass this as an indicator of that since we cant pass the optional value. All of the methods have this construct for consistency among them. I am aware that 1-2 of them can avoid using it and return List<String> but have left for the following change wit the indicator. The following change with the return value indicator that I left a TODO for will hopefully remove the need for it and I will remove all of them